Entrypoint detection stages

TyOverby opened the pull request.

TyOverby commented:
Ready for initial review @jaredpar @VSadov

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,94 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                var intOrVoidEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
+                // Validity and diagnostics are also tracked because they must be conditionally handled 
+                // if there are not any "traditional" entrypoints found.
+                var taskEntryPoints = ArrayBuilder<(bool IsValid, MethodSymbol Candidate, DiagnosticBag SpecificDiagnostics)>.GetInstance();
```
`taskEntryPoints` is not freed.

TyOverby replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,94 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                 foreach (var candidate in entryPointCandidates)
                 {
-                    if (!HasEntryPointSignature(candidate, warnings))
+                    var perCandidateBag = DiagnosticBag.GetInstance();
```
`perCandidateBag` is not freed.

TyOverby replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1580,45 +1633,45 @@ private bool HasEntryPointSignature(MethodSymbol method, DiagnosticBag bag)
             if (!returnsTaskOrTaskOfInt && method.IsAsync && this.LanguageVersion >= LanguageVersion.CSharp7_1)
             {
-                return false;
+                return (false, returnsTaskOrTaskOfInt);
```
Are we ever using `IsTaskLike` if `IsCandidate` is `false`? Consider returning a single value rather than two separate `bool` values. Perhaps `enum EntryPointKind { NotCandidate, VoidOrInt, TaskOrTaskOfInt }`.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,100 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                    else
+                    {
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
+                        noMainFoundDiagnostics.AddRange(perCandidateBag);
```
`perCandidateBag` is not freed.

TyOverby replied:
Done.

jaredpar commented:
CC @dotnet/roslyn-compiler 

Don't think the edit of an existing comment to add that alias produces new emails.  At least I didn't see one.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,100 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.IsEmpty())
```
Does `IsEmpty` require an `IEnumerable`? If so, consider using `.Count == 0` instead.

TyOverby replied:
Done.

TyOverby commented:
@cston: This PR implements a discussion that we had earlier about splitting the found entrypoints into different collections.  Thanks for taking the time to review it!

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.Count() == 0)
```
`Count` is a property. Is this binding to an `IEnumerable` extension method instead.

TyOverby replied:
Oh yep, you're right.

TyOverby replied:
Done.

cston commented:
>         internal bool ReturnsAwaitableToVoidOrInt(MethodSymbol method, DiagnosticBag diagnostics)

`private`?

---
Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:1594 in ee1a204. [](commit_id = ee1a2043aeb29ce9b8c26401fafe577dd283d3e0, deletion_comment = False)

TyOverby commented:
@dotnet-bot test ubuntu_16_debug_prtest please

TyOverby commented:
@cston: I don't know what code review tool you are using, but I can't "reply" to your comment on `internal bool ReturnsAwaitableToVoidOrInt`.  The reason that it is internal instead of private is that this method is also used by the emitter.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
+                        noMainFoundDiagnostics.AddRange(perCandidateBag);
+                        perCandidateBag.Free();
                     }
```
Perhaps check `IsTaskLike` first. That would simplify the `IsCandidate` check and `perCandidateBag` could be freed in one place rather than two.
```
if (IsTaskLike)
{
    taskEntryPoints.Add(...);
}
else
{
    if (IsCandidate)
    {
        intOrVoidEntryPoints.Add(candidate);
    }
    else
    {
        noMainFoundDiagnostics.Add(...);
    }
    perCandidatesBag.Free();
}

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -5074,4 +5074,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   </data>
-</root>
+  <data name="ERR_NonTaskMainCantBeAsync" xml:space="preserve">
+    <value>Async Main methods must return Task or Task&lt;int&gt;</value>
```
>Task&lt;int&gt; [](start = 50, length = 15)

Does it have to be ```Task<int>``` though?

TyOverby replied:
Not technically, but I think this is fine for an error message.

I'd be willing to change it, but "Async Main methods must return either Task or Task<T> whereby the return type of GetAwaiter().GetResult() is either `void` or `int`" seems a bit verbose for an error message.

AlekseyTs replied:
I think we can simply say that int or void returning Main cannot be async.

VSadov replied:
+1 for  "void returning Main cannot be async"

---
In reply to: [114608904](https://github.com/dotnet/roslyn/pull/19093#discussion_r114608904) [](ancestors = 114608904)

TyOverby replied:
Changed.

---
In reply to: [114619543](https://github.com/dotnet/roslyn/pull/19093#discussion_r114619543) [](ancestors = 114619543,114608904)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -3,13 +3,14 @@
     {
-
         [Fact]
+        [CompilerTrait(CompilerFeature.AsyncMain)]
```
>[CompilerTrait(CompilerFeature.AsyncMain)] [](start = 8, length = 42)

I think this can simply be applied  to the test type instead.

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/MethodTests.cs:
```diff
@@ -252,12 +251,12 @@ public partial class A {
-            var model = comp.GetSemanticModel(tree);
-            var token = tree.GetRoot().DescendantTokens().Where(t => t.Kind() == SyntaxKind.VoidKeyword).Skip(1).First();
-            var node = tree.GetRoot().FindNode(token.Span);
+            var node = tree.GetRoot().DescendantNodes().OfType<PredefinedTypeSyntax>().Where(n => n.Keyword.Kind() == SyntaxKind.VoidKeyword).ToList()[1];
```
>ToList()[1] [](start = 142, length = 11)

```Last()```, or ```ElementAt(1)```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
-                        warnings.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
-                        continue;
+                        intOrVoidEntryPoints.Add(candidate);
+                        perCandidateBag.Free();
```
>perCandidateBag.Free(); [](start = 24, length = 23)

Consider asserting that the bag is empty.

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1479,5 +1479,6 @@ internal enum ErrorCode
         ERR_BadDynamicMethodArgDefaultLiteral = 9000,
         ERR_DefaultLiteralNotValid = 9001,
         WRN_DefaultInSwitch = 9002,
+        ERR_NonTaskMainCantBeAsync = 9003,
```
>ERR_NonTaskMainCantBeAsync [](start = 8, length = 26)

Can we simply reuse ERR_MainCantBeAsync slot?

TyOverby replied:
I'm reusing the old number, but with a new name and message.

AlekseyTs commented:
>   <data name="ERR_MainCantBeAsync" xml:space="preserve">

It looks like this entry should be removed.

---
Refers to: src/Compilers/CSharp/Portable/CSharpResources.resx:3599 in ee1a204. [](commit_id = ee1a2043aeb29ce9b8c26401fafe577dd283d3e0, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                            continue;
+                        }
+
+                        if (!IsValid)
```
>if (!IsValid) [](start = 24, length = 13)

It feels like this ```if``` should be first in this loop (for consistency with int/void entry points).

cston replied:
Can these two `if` blocks be extracted to a helper method used in both the non-task-like and task-like candidates? (It would mean performing both checks for non-task-like candidates in the same loop, perhaps the first `foreach`, so the candidate is never added to `intOrVoidEntryPoints`.)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                    // We can't add those Errors to the general diagnostics bag because it would break previously-working programs.
+                    // The fact that these warnings are not added when csc is invoked with /main is possibly a bug, and is tracked at
+                    // https://github.com/dotnet/roslyn/issues/18964
+                    diagnostics.AddRange(noMainFoundDiagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Warning));
```
>noMainFoundDiagnostics.AsEnumerable().Where(d => d.Severity == DiagnosticSeverity.Warning) [](start = 41, length = 90)

Linq is banned in compilers. Also, if the point of this code is to preserve old behavior, then we should probably be more precise and add only  WRN_MainCantBeGeneric and WRN_InvalidMainSig warnings.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -819,11 +1014,12 @@ class A
-    }").WithArguments("async main", "7.1").WithLocation(6, 5)
-                );
+    }").WithArguments("async main", "7.1").WithLocation(6, 5),
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
```
>Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1) [](start = 16, length = 57)

Please add corresponding message as a comment.

TyOverby replied:
Done

TyOverby replied:
done.

AlekseyTs commented:
Done with review pass.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1565,11 +1628,11 @@ internal bool ReturnsAwaitableToVoidOrInt(MethodSymbol method, DiagnosticBag dia
         /// or <see cref="System.Threading.Tasks.Task{T}" /> where T is an int.
```
>where T is an int [](start = 61, length = 17)

It looks like this comment is no longer accurate.

TyOverby commented:
@AlekseyTs Your comment 

> It looks like this entry should be removed.

Isn't attached to any code, I don't know what it is that you think can be removed.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,120 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+
+                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
 
+                bool CheckGenericity(MethodSymbol candidate)
```
Consider including `IsValid` check as well (to report `ErrorCode.WRN_InvalidMainSig`) since that same check appears in both cases where this local function is called.

TyOverby replied:
But `IsValid` is always true in the `intOrVoidEntryPoints` case.  (It had to be for the member to end up in the list)

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,114 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                // there were not any main methods found.
+                DiagnosticBag noMainFoundDiagnostics = DiagnosticBag.GetInstance();
+
+                bool CheckGenericity(MethodSymbol candidate, bool isValid, DiagnosticBag specificDiangostics) 
```
Typo: `specificDiagnostics`

cston replied:
Method name is stale. Perhaps: `CheckSignature`.

TyOverby replied:
Fixed

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,114 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                    {
+                        if (!CheckGenericity(candidate, IsCandidate, perCandidateBag))
+                        {
+                            continue;
```
`continue` is not needed since we need to call `perCandidateBag.Free()` in this case.

TyOverby replied:
Done.  I've also renamed the local function to IsValid.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                        continue;
+                    }
                     viableEntryPoints.Add(candidate);
                 }
```
Could the body of this `foreach` be inlined in the previous `foreach`, when `CheckValid()` returns true? Then, `intOrVoidEntryPoints` can be removed.

TyOverby replied:
Done.

AlekseyTs commented:
@TyOverby Please ping when the PR is ready for another round of reviews.

TyOverby commented:
@AlekseyTs ping

jaredpar commented:
>   <data name="ERR_MainCantBeAsync" xml:space="preserve">

Still see this.  Agree it can be removed. 

---
In reply to: [298794787](https://github.com/dotnet/roslyn/pull/19093#issuecomment-298794787) [](ancestors = 298794787)

---
Refers to: src/Compilers/CSharp/Portable/CSharpResources.resx:3599 in ee1a204. [](commit_id = ee1a2043aeb29ce9b8c26401fafe577dd283d3e0, deletion_comment = False)

jaredpar commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1104,7 +1104,7 @@ internal enum ErrorCode
         ERR_VarargsAsync = 4006,
         ERR_ByRefTypeAndAwait = 4007,
         ERR_BadAwaitArgVoidCall = 4008,
-        // ERR_MainCantBeAsync = 4009,
```
>  // ERR_MainCantBeAsync = 400 [](start = 6, length = 30)

Why comment out instead of delete?

TyOverby replied:
Seemed like the convention.  Should it be removed?

jaredpar replied:
The convention, at least as well as I can surmise, is to comment out when there is a gap in the ordering.  For instance ERR_BadAwaitWithoutAsyncAnonMeth (20-ish lines down).  It was removed but the error slot 4035 still exists.  In this case we're replacing an error hence 4009 is not a gap anymore.  

This convention makes it easy to find places to put new errors.  Vs having to scan through te file looking for subtle changes in the number. 

---
In reply to: [114816031](https://github.com/dotnet/roslyn/pull/19093#discussion_r114816031) [](ancestors = 114816031)

TyOverby commented:
@jaredpar, @AlekseyTs 

This comment
> It looks like this entry should be removed
>
> Refers to: src/Compilers/CSharp/Portable/CSharpResources.resx:3599 in ee1a204. [](commit_id = ee1a204, deletion_comment = False)

looks out of date. At least, I don't see anything at 3599 that needs removal anymore

jaredpar commented:
> looks out of date. At least, I don't see anything at 3599 that needs removal anymore

Yep.  I saw that comment and it was decidedly attached to a red piece of code.  But I could still read it and thought "why hasn't it been deleted?".  Then a few seconds later I realized "it's red because it's deleted" and a I deleted the comment.  

Another lesson in @jaredpar should be quiet until he finishes his morning coffee.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                {
+                    foreach (var (IsValid, Candidate, SpecificDiagnostics) in taskEntryPoints)
+                    {
+                        if (!CheckValid(Candidate, IsValid, SpecificDiagnostics))
```
>if (!CheckValid(Candidate, IsValid, SpecificDiagnostics)) [](start = 24, length = 57)

It looks like this condition can be inverted and simply combined with condition in the next ```if```.

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                        {
+                            intOrVoidEntryPoints.Add(candidate);
+                        }
+                        perCandidateBag.Free();
```
>perCandidateBag [](start = 24, length = 15)

Please assert the bag is empty.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,105 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                    {
+                        if (!CheckValid(candidate, IsCandidate, perCandidateBag))
+                        {
+                            continue;
```
`perCandidateBag` is not freed in this `if` or `if` below. Perhaps use `if (...) { ... } else if (...) { ... } else { ... } perCandidateBag.Free();`.

AlekseyTs replied:
I think the code can be rewritten without ```continue```.
Something like:
```
if (CheckValid)
{
    if (isAsync)
    { report an error }
    else
    { add }
}

free the bag.
```

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                    foreach (var diagnostic in noMainFoundDiagnostics.AsEnumerable())
+                    {
+                        if (diagnostic.Severity == DiagnosticSeverity.Error) {
+                            diagnostics.Add(diagnostic.WithSeverity(DiagnosticSeverity.Warning));
```
>diagnostics.Add(diagnostic.WithSeverity(DiagnosticSeverity.Warning)); [](start = 28, length = 69)

I don't think changing severity on the diagnostics is the right thing to do. Also, why, for example, would we want to report an obsolete diagnostics associated with GetAwaiter for an entry point that isn't going to be used? Please add a test for this scenario. I still think that we should be reporting only specific warnings, which are reported in CheckValid.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/Source/MethodTests.cs:
```diff
@@ -252,12 +251,12 @@ public partial class A {
-            var model = comp.GetSemanticModel(tree);
-            var token = tree.GetRoot().DescendantTokens().Where(t => t.Kind() == SyntaxKind.VoidKeyword).Skip(1).First();
-            var node = tree.GetRoot().FindNode(token.Span);
+            var node = tree.GetRoot().DescendantNodes().OfType<PredefinedTypeSyntax>().Where(n => n.Keyword.Kind() == SyntaxKind.VoidKeyword).ToList().Last();
```
>ToList() [](start = 142, length = 8)

Why do you still do ToList?

TyOverby replied:
Fixed

AlekseyTs commented:
>         public void AsyncEmitMainOfIntTest()

Could you please clarify why this test is removed?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:466 in ee1a204. [](commit_id = ee1a2043aeb29ce9b8c26401fafe577dd283d3e0, deletion_comment = True)

AlekseyTs commented:
>         public void MainCantBeAsyncVoid()

Could you please clarify why this test is removed?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:786 in ee1a204. [](commit_id = ee1a2043aeb29ce9b8c26401fafe577dd283d3e0, deletion_comment = True)

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -5074,4 +5071,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>A tuple may not contain a value of type 'void'.</value>
   </data>
+  <data name="ERR_NonTaskMainCantBeAsync" xml:space="preserve">
+    <value>A void or int returning Main cannot be async</value>
```
>Main [](start = 35, length = 4)

It is probably better to use term "entry point" instead of "Main". It was used by the original message and is more precise because, in general, void or int returning methods named Main can be async.

TyOverby replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +1004,156 @@ class A
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1));
```
>Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1) [](start = 16, length = 57)

Please add comments as for other diagnostics base-lines.

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +1004,156 @@ class A
+        System.Console.WriteLine(""Async Void Main"");
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1)).VerifyDiagnostics();
```
> var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1)).VerifyDiagnostics(); [](start = 11, length = 194)

It is not clear why we starting allowing async void entry point. Also, success scenarios should be executed and runtime behavior should be verified.

TyOverby replied:
This is an incomplete test that I probably shouldn't have committed (notice that the [fact] attribute is missing.

I've finished writing the test.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +1004,156 @@ class A
+        System.Console.WriteLine(""Async Void Main"");
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
```
>var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe).VerifyDiagnostics( [](start = 12, length = 109)

If you expect the behavior to change for the next language version, consider explicitly specifying the version that you are testing. Otherwise, the test is going to break once we bump the default version. See if other tests have the same issue as well.

TyOverby replied:
Done for all tests.

AlekseyTs commented:
Done with review pass.

TyOverby commented:
>  think the code can be rewritten without continue.
> Something like:
> 
> if (CheckValid)
> {
>     if (isAsync)
>     { report an error }
>     else
>     { add }
> }

I like this. Done

TyOverby commented:
> Could you please clarify why this test is removed?

Turns both of these were duplicate tests.

TyOverby commented:
> It is probably better to use term "entry point" instead of "Main". It was used by the original message and is more precise because, in general, void or int returning methods named Main can be async.

Done.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                                viableEntryPoints.Add(candidate);
+                            }
+                        }
+                        perCandidateBag.Free();
```
Assert `perCandidateBag` is empty inside `if (CheckValid()) { ... }`.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                            }
+                            else
+                            {
+                                viableEntryPoints.Add(candidate);
```
Assert `perCandidateBag` is empty.

cston replied:
Or alternatively, add `diagnostics.AddRange(perCandidateBag);`.

TyOverby replied:
^ I like `diagnostics.AddRange(perCandidateBag);`, I'm going with that one.

cston commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -1561,15 +1623,16 @@ internal bool ReturnsAwaitableToVoidOrInt(MethodSymbol method, DiagnosticBag dia
-        /// - the return type is either void, int, <see cref="System.Threading.Tasks.Task" />,
-        /// or <see cref="System.Threading.Tasks.Task{T}" /> where T is an int.
+        /// - the return type is either void, int, or returns a <see cref="System.Threading.Tasks.Task" />,
+        /// or <see cref="System.Threading.Tasks.Task{T}" /> where the resturn type of GetAwaiter().GetResult() 
```
Typo: `return type`

TyOverby replied:
Thanks, Done.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,173 @@ class A
+    {
+        await Task.Factory.StartNew(() => { });
+        System.Console.WriteLine(""Async Void Main"");
+    }
```
For completeness, consider including an `async static int Main() { ... }` here.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,173 @@ class A
+                // (10,30): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point
+                //     async static Task<float> Main(string[] args)
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("A.Main(string[])").WithLocation(11, 30));
+            CompileAndVerify(compilation, expectedOutput: "Non Task Main", expectedReturnCode: 0);
```
It seems surprising that the following is allowed, where `static void Main()` is the entry point:
```
static void Main() { ... }
static Task<int> Main(string[] args) { ... }
```
but the following results in an error:
```
static void Main() { ... }
static Task<float> Main(string[] args) { ... }
```
Why report an error for `static Task<float> Main()` if we don't plan on using it as the entry point?

TyOverby replied:
Yeah, I agree, but my current implementation is analogous to the previous non-task-based handling of mains.

```c#
static int main(){}
// WARNING
static float main(string[] args) {}
```

```c#
static int main(){}
// WARNING
static Task<float> main(string[] args) {}
```

```c#
static Task<int> main(){}
// WARNING
static Task<float> main(string[] args) {}
```

TyOverby commented:
ping @AlekseyTs for another review.

cston commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -880,7 +976,7 @@ class A
     }
 }";
-            CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
+            CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics(
```
>, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7) [](start = 83, length = 80)

If we don't expect the behavior to be version specific, we probably shouldn't use specific version for this test. Perhaps we should test it against default version and the latest version, latest will give us coverage for 7.1.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
     }
 }";
-            CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe).VerifyDiagnostics(
+            CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics(
```
>, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7) [](start = 83, length = 80)

Same comment as for the previous test.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
+        System.Console.WriteLine(""Task Main"");
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe.WithMainTypeName("A"), parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics();
```
>, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)) [](start = 123, length = 81)

If we don't expect the behavior to be version specific, we probably shouldn't use specific version for this test. Perhaps we should test it against default version and the latest version, latest will give us coverage for 7.1.

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
+        }
+
+        [Fact]
+        public void TaskIntAndTaskFloat_CSharp7()
```
>CSharp7 [](start = 40, length = 7)

It looks like the test is actually using 7.1

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
+        return 0.0f;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics(
```
>, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7) [](start = 101, length = 80)

If we don't expect the behavior to be version specific, we probably shouldn't use specific version for this test. Perhaps we should test it against default version and the latest version, latest will give us coverage for 7.1.

TyOverby replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
+        return 0.0f;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe.WithMainTypeName("A"), parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics();
```
>, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7) [](start = 123, length = 80)

If we don't expect the behavior to be version specific, we probably shouldn't use specific version for this test. Perhaps we should test it against default version and the latest version, latest will give us coverage for 7.1.

TyOverby replied:
done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -901,12 +997,207 @@ class A
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1)).VerifyDiagnostics(
+                // (6,23): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point
```
>// (6,23): warning CS0028: 'A.Main(string[])' has the wrong signature to be an entry point [](start = 16, length = 90)

Why don't we report the same error as in the previous test?

AlekseyTs commented:
Done with review pass.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -52,21 +53,115 @@ static class Program {
+                Diagnostic(ErrorCode.WRN_SameFullNameThisAggAgg, "Task<int>").WithArguments("", "System.Threading.Tasks.Task<T>", "mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089", "System.Threading.Tasks.Task<TResult>"),
+                // (11,22): warning CS0028: 'Program.Main()' has the wrong signature to be an entry point
+                //     static Task<int> Main() {
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()"));
```
>Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()")); [](start = 16, length = 82)

It looks like we lost locations in the base-line again. Please restore and check other tests too.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -657,14 +746,14 @@ class A
             compilation.VerifyDiagnostics(
-                // (6,22): error CS1983: The return type of an async method must be void, Task or Task<T>
+// (6,22): error CS1983: The return type of an async method must be void, Task or Task<T>
                 //     async static int Main()
```
It looks like formatting is off.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +998,220 @@ class A
+        return 1;
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7)).VerifyDiagnostics(
```
>parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7) [](start = 103, length = 78)

Do you expect the behavior to change in the next version?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +998,220 @@ class A
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.ReleaseExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1)).VerifyDiagnostics(
+// (6,23): error CS4009: A void or int returning entry point cannot be async
```
It looks like formatting is off.

AlekseyTs commented:
Done with review pass.

TyOverby commented:
@AlekseyTs Ping for another review

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -52,21 +53,115 @@ static class Program {
-                // (11,12): error CS1986: 'await' requires that the type Task<int> have a suitable GetAwaiter method
-                //     static Task<int> Main() {
-                Diagnostic(ErrorCode.ERR_BadAwaitArg, "Task<int>").WithArguments("System.Threading.Tasks.Task<int>").WithLocation(11, 12),
+                Diagnostic(ErrorCode.WRN_SameFullNameThisAggAgg, "Task<int>").WithArguments("", "System.Threading.Tasks.Task<T>", "mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089", "System.Threading.Tasks.Task<TResult>"),
```
>Diagnostic(ErrorCode.WRN_SameFullNameThisAggAgg, "Task<int>").WithArguments("", "System.Threading.Tasks.Task<T>", "mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089", "System.Threading.Tasks.Task<TResult>"), [](start = 16, length = 233)

It looks like we lost locations in the base-line again. Please restore and check other tests too.

AlekseyTs commented:
Done with review pass.

TyOverby commented:
@AlekseyTs done

jaredpar commented:
retest windows_debug_unit64_prtest please

TyOverby commented:
ping @AlekseyTs

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -52,21 +53,114 @@ static class Program {
+                Diagnostic(ErrorCode.WRN_SameFullNameThisAggAgg, "Task<int>").WithArguments("", "System.Threading.Tasks.Task<T>", "mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089", "System.Threading.Tasks.Task<TResult>"),
+                // (11,22): warning CS0028: 'Program.Main()' has the wrong signature to be an entry point
+                //     static Task<int> Main() {
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()"));
```
>Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()")) [](start = 16, length = 81)

Location is not verified.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -548,6 +658,8 @@ class A
+                Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1),
                 // (6,5): error CS8107: Feature 'async main' is not available in C# 7. Please use language version 7.1 or greater.
                 //     async static Task Main(string[] args)
                 Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task Main(string[] args)
```
> Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task Main(string[] args) [](start = 15, length = 96)

Location is not verified.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -751,13 +836,14 @@ class A
+            Diagnostic(ErrorCode.ERR_NoEntryPoint).WithLocation(1, 1),
             // (6,5): error CS8107: Feature 'async main' is not available in C# 7. Please use language version 7.1 or greater.
             //     async static Task<int> Main(string[] args)
             Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task<int> Main(string[] args)
```
>Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task<int> Main(string[] args) [](start = 12, length = 100)

Location is not verified.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs:
```diff
@@ -908,5 +996,220 @@ class A
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "Main").WithLocation(12, 30),
+                // (6,5): error CS8107: Feature 'async main' is not available in C# 7. Please use language version 7.1 or greater.
+                //     async static Task<int> Main()
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task<int> Main()
```
>Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, @"async static Task<int> Main() [](start = 16, length = 87)

Location is not verified.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs:
```diff
@@ -3434,6 +3434,8 @@ class Test
                 //     async public static Task Main()
                 Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "Main"),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint),
```
>Diagnostic(ErrorCode.ERR_NoEntryPoint), [](start = 16, length = 39)

Location is not verified. And above as well.

AlekseyTs commented:
Done with review pass.

TyOverby commented:
@AlekseyTs I temporarily made it so that missing WithLocation was a test failure, then I fixed all the failing tests.  Assuming that my failure-inducing code was correct, I've cleaned up the remainder of the missing WithLocation errors.
